### PR TITLE
Support retry on truncation

### PR DESCRIPTION
**Pull request by folbricht**
Introduces `truncate-retry` group to support UDP truncation. Also improves handling of the TC bit when using UDP listeners as per #170

### CODE REVIEW

**Comment by folbricht** [comment]
Nice, didn't notice that one. Using this now, thanks.

File: cache_test.go [hunk 1]
```diff
@@ -173,3 +173,28 @@ func TestRoundRobinShuffle(t *testing.T) {
...
+			a.SetReply(q)
+			a.Truncated = true
+			return a, nil
+		},
+	}
+
+	c := NewCache("test-cache", r, CacheOptions{})
+
+	// Both queries should hit the upstream resolver
+	q.SetQuestion("test.com.", dns.TypeA)
```
Comment (charlieporth1) [inline cache_test.go:193]: This domain `test.com` worries me unless you own the domain.
I would either query `www.google.com` or `github.com` for long term change and whitelist reasons. Those domains are much more "stable" as of today than `test.com` which I believe will exchange hands and possibly go `NXDOMAIN`, or `SERVFAIL` if push comes to shove. To add to that in my config I use pihole as the root lightweight blocker. If someone has `test.com` in there  base level DNS then that could add an issue as well depeneding on there config and setup if its blocked or not.

Comment (folbricht) [reply cache_test.go]: Good point. While these queries don't actually hit any public resolvers, it's not good practice in general. I switched to using example.com instead which is intended to be used for testing.

Comment (charlieporth1) [reply cache_test.go]: I just checked test.com is blocked on my config

**Review by charlieporth1 — APPROVED** [review]

**Pull request merged**